### PR TITLE
WiP: Double Publish and double Marry Malfeasance Proofs for ATX v2

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation

This adds the first two malfeasance proofs for ATXv2: double publish and double marry.

### Double Publish
If an identity publishes two ATXs with the same publish epoch they become malfeasant. This proof verifies that a given ATX was created by the same identity and targets the same epoch. It works basically the same as the previous double publish proof, except that merkle trees are used instead of a signature where the payloads first 4 bytes are the publish epoch.

### Double Marry
If an identity is part of more than one marriage it is considered malfeasant. This Proof shows that two ATXs containing marriage certificates contain certificates signed by the same identity.

To simplify this proof the main identity (the one that signs the ATX) has to include a marriage certificate marrying themselves, otherwise we would need additional proofs: i.e. 2 ATXs signed by the same identity containing marriage certificates and one for the case where one ATX containing marriage certificates is signed by identity A and another ATX signed by B contains a marriage certificate of A.

### Creating and verifying proofs
Both types of proofs have a `New...Proof` function that creates them. This function does some basic checks to prevent creating an invalid malfeasance proof if two ATXs that do not show malfeasant behavior are passed as arguments.

The wire types for the malfeasance proofs have a `Valid` method that can be called to check if the malfeasance proof is valid. It returns an `error` that is not `nil` when the proof is invalid and explains why it is invalid.

## Test Plan

For both malfeasance proofs multiple tests have been added to verify they are working correctly.

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: activation/wire/malfeasance.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type ATXProof struct {
+	// ProofType is the type of proof that is being provided.
+	ProofType ProofType
+	// Certificates is a slice of marriage certificates showing which identities belong to the same marriage set as
+	// the one proven to be malfeasant. Up to 1024 can be put into a single proof, since by repeatedly marrying other
+	// identities there can be much more than 256 in a malfeasant marriage set. Beyond that a second proof could be
+	// provided to show that additional identities are part of the same malfeasant marriage set.
+	Certificates []ProofCertificate `scale:"max=1024"`
```
Comment (poszu) [inline activation/wire/malfeasance.go:25]: Maybe name it 
```suggestion
	Certificates []MarriageCertificate `scale:"max=1024"`
```

File: activation/wire/malfeasance.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// ProofCertificate proofs that two identities are part of the same marriage set.
+type ProofCertificate struct {
+	// Target of the certificate, i.e. the identity that signed the ATX containing the original certificate.
+	Target types.NodeID
+	// ID is the identity that signed the certificate.
+	ID types.NodeID
+	// Signature is the signature of the certificate, i.e. ID signed with the key of SmesherID.
+	Signature types.EdSignature
+}
```
Comment (poszu) [inline activation/wire/malfeasance.go:38]: Having `Target` in each certificate is quite redundant as in most cases, all certs will share it. How about structure like this:
```go
type MarriageCertificates {
	Target types.NodeID
	Certificates []MarriageCertificate `scale:"max=256"`
}

type ATXProof struct {
	ProofType ProofType
	Certificates []MarriageCertificates `scale:"max=2"`
	Proof []byte `scale:"max=1048576"`
}
```

File: activation/wire/malfeasance_double_marry.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+				NodeID: nodeID,
+
+				MarriageRoot:  types.Hash32(atx1.Marriages.Root()),
+				MarriageProof: proof1,
+
+				CertificateSignature: atx1.Marriages[atx1Index].Signature,
+				CertificateIndex:     uint64(atx1Index),
+				CertificateProof:     certProof1,
+
+				SmesherID: atx1.SmesherID,
```
Comment (poszu) [inline activation/wire/malfeasance_double_marry.go]: It's super easy to confuse variables here. How about creating a dedicated function to create a `MarryProof`?

```go
func createMarryProof(atx *ActivationTxV2, id types.NodeID) (MarryProof, error) {...}

proof1, err := createMarryProof(atx1, nodeID)
if err != nil {...}
proof2, err := createMarryProof(atx2, nodeID)
if err != nil {...}
proof := &ProofDoubleMarry{Proofs: [2]MarryProof{proof1, proof2}}

File: activation/wire/malfeasance.go (hunk 1 above)
Comment (fasmat) [reply activation/wire/malfeasance.go:25]: `MarriageCertificate` already exists as a type in the `wire` package (for v2 ATXs) but this is a different type it not only contains the signature and the payload but also the signer (which isn't needed for ATXs).

File: activation/wire/malfeasance_double_marry.go (hunk 3 above)
Comment (fasmat) [reply activation/wire/malfeasance_double_marry.go]: I refactored this section as suggested

**Comment by fasmat** [comment]
superseded by https://github.com/spacemeshos/go-spacemesh/pull/6043

**Pull request closed without merging**